README.txt
Welcome to FlaskFit, an example of creating a
web-based curve and surface fitting tool with
Python Flask and the pyeq3 fitting library.
see http://commonproblems.readthedocs.io/en/latest/


Step 1: Install flask, scipy and matplotlib

On Debian or Ubuntu Linux, you can use this command:

sudo apt-get install python3-flask python3-scipy python3-matplotlib


On other operating systems, try the Canopy Express Free version:

https://store.enthought.com/

and install Flask with:

easy_install Flask


Finally, install the pyeq3 fitting library using pip:

pip3 install pyeq3



Step 2: Run the FlaskFit example

From a command prompt run this command:

python3 FlaskFit.py

The command prompt should then display:

* Running on http://0.0.0.0:5000/

When you open this URL you should see a form
that allows curve and surface fitting, with graphs.
The address "0.0.0.0" allows port 5000 to be seen
externally.  On my development laptop, I can open
the URL "127.0.0.1:5000/" and see the web server.

Use Control-C to exit the Flask development server.



Step 3: Celebrate

You are now curve and surface fitting data using Flask!



If you would like to create a feature complete web site
with django, see https://github.com/zunzun/zunzunsite3
which has "function finding" and animated surface rotations
along with SMP parallel processing for multi-core servers.


CHANGELOG
The change log can be viewed at:

https://github.com/zunzun/FlaskFit/commits/master


TextUtils.py
import os, sys
import scipy

import pyeq3


def DataArrayStatistics(inArray):
    returnString = '' # build this as we progress
    
    # must at least have max and min
    minData = min(inArray)
    maxData = max(inArray)
    
    if maxData == minData:
        returnString += 'All data has the same value,\n'
        returnString += "value = %-.16E\n" % (minData)
        returnString += 'statistics cannot be calculated.'
    else:
        returnString += "max = %-.16E\n" % (maxData)
        returnString += "min = %-.16E\n" % (minData)
        
        try:
            temp = scipy.mean(inArray)
            returnString += "mean = %-.16E\n" % (temp)
        except:
            returnString += "mean gave error in calculation\n"

        try:
            temp = scipy.stats.sem(inArray)
            returnString += "standard error of mean = %-.16E\n" % (temp)
        except:
            returnString += "standard error of mean gave error in calculation\n"

        try:
            temp = scipy.median(inArray)
            returnString += "median = %-.16E\n" % (temp)
        except:
            returnString += "median gave error in calculation\n"

        try:
            temp = scipy.var(inArray)
            returnString += "variance = %-.16E\n" % (temp)
        except:
            returnString += "variance gave error in calculation\n"

        try:
            temp = scipy.std(inArray)
            returnString += "std. deviation = %-.16E\n" % (temp)
        except:
            returnString += "std. deviation gave error in calculation\n"

        try:
            temp = scipy.stats.skew(inArray)
            returnString += "skew = %-.16E\n" % (temp)
        except:
            returnString += "skew gave error in calculation\n"

        try:
            temp = scipy.stats.kurtosis(inArray)
            returnString += "kurtosis = %-.16E\n" % (temp)
        except:
            returnString += "kurtosis gave error in calculation\n"
    
    return returnString


def SaveCoefficientAndFitStatistics(in_filePathFitStatistics, in_equation):
    
    outputFile = open(in_filePathFitStatistics, 'w')
    
    if in_equation.upperCoefficientBounds or in_equation.lowerCoefficientBounds:
        outputFile.write('This model has coefficient bounds. Parameter statistics may\n')
        outputFile.write('not be valid for parameter values at or near the bounds.\n')
        outputFile.write('\n')
    
    outputFile.write('Degress of freedom error ' + str(in_equation.df_e) + '\n')
    outputFile.write('Degress of freedom regression ' + str(in_equation.df_r) + '\n')
    
    if in_equation.rmse == None:
        outputFile.write('Root Mean Squared Error (RMSE): n/a\n')
    else:
        outputFile.write('Root Mean Squared Error (RMSE): ' + str(in_equation.rmse) + '\n')
    
    if in_equation.r2 == None:
        outputFile.write('R-squared: n/a\n')
    else:
        outputFile.write('R-squared: ' + str(in_equation.r2) + '\n')
    
    if in_equation.r2adj == None:
        outputFile.write('R-squared adjusted: n/a\n')
    else:
        outputFile.write('R-squared adjusted: ' + str(in_equation.r2adj) + '\n')
    
    if in_equation.Fstat == None:
        outputFile.write('Model F-statistic: n/a\n')
    else:
        outputFile.write('Model F-statistic: ' + str(in_equation.Fstat) + '\n')
    
    if in_equation.Fpv == None:
        outputFile.write('Model F-statistic p-value: n/a\n')
    else:
        outputFile.write('Model F-statistic p-value: ' + str(in_equation.Fpv) + '\n')
    
    if in_equation.ll == None:
        outputFile.write('Model log-likelihood: n/a\n')
    else:
        outputFile.write('Model log-likelihood: ' + str(in_equation.ll) + '\n')
    
    if in_equation.aic == None:
        outputFile.write('Model AIC: n/a\n')
    else:
        outputFile.write('Model AIC: ' + str(in_equation.aic) + '\n')
    
    if in_equation.bic == None:
        outputFile.write('Model BIC: n/a\n')
    else:
        outputFile.write('Model BIC: ' + str(in_equation.bic) + '\n')
    
    outputFile.write('\n')
    outputFile.write("Individual Parameter Statistics:\n")
    for i in range(len(in_equation.solvedCoefficients)):
        if type(in_equation.tstat_beta) == type(None):
            tstat = 'n/a'
        else:
            tstat = '%-.5E' %  ( in_equation.tstat_beta[i])
    
        if type(in_equation.pstat_beta) == type(None):
            pstat = 'n/a'
        else:
            pstat = '%-.5E' %  ( in_equation.pstat_beta[i])
    
        if type(in_equation.sd_beta) != type(None):
            outputFile.write("Coefficient %s = %-.16E, std error: %-.5E\n" % (in_equation.GetCoefficientDesignators()[i], in_equation.solvedCoefficients[i], in_equation.sd_beta[i]))
        else:
            outputFile.write("Coefficient %s = %-.16E, std error: n/a\n" % (in_equation.GetCoefficientDesignators()[i], in_equation.solvedCoefficients[i]))
        outputFile.write("          t-stat: %s, p-stat: %s, 95 percent confidence intervals: [%-.5E, %-.5E]\n" % (tstat,  pstat, in_equation.ci[i][0], in_equation.ci[i][1]))
        
    outputFile.write('\n')
    outputFile.write("Coefficient Covariance Matrix:\n")
    for i in  in_equation.cov_beta:
        outputFile.write(str(i) + '\n')
    
    # absolute error statistics
    outputFile.write('\n\n\n')
    outputFile.write('Absolute Error Statistics:\n')
    outputFile.write(DataArrayStatistics(in_equation.modelAbsoluteError))
    
    if in_equation.dataCache.DependentDataContainsZeroFlag == 1:
        outputFile.write('\n\n\n')
        outputFile.write('Percent Error Statistics cannot be calculated, as\ndependent data contains at least one value of exactly zero.\n')
    else:
        outputFile.write('\n\n\n')
        outputFile.write('Percent Error Statistics:\n')
        outputFile.write(DataArrayStatistics(in_equation.modelPercentError))
        
    outputFile.close()


def SaveSourceCode(in_sourceCodeFilePath,  in_equation):
    
    outputFile = open(in_sourceCodeFilePath, 'w')
        
    outputFile.write('<html><body>\n\n')
    
    try:
        outputFile.write('<b>C++</b><br><textarea rows="20" cols="85" wrap="OFF">')
        outputFile.write(pyeq3.outputSourceCodeService().GetOutputSourceCodeCPP(in_equation))
        outputFile.write('</textarea><br><br>\n\n')
    except:
        pass

    try:
        outputFile.write('<b>CSHARP</b><br><textarea rows="20" cols="85" wrap="OFF">')
        outputFile.write(pyeq3.outputSourceCodeService().GetOutputSourceCodeCSHARP(in_equation))
        outputFile.write('</textarea><br><br>\n\n')
    except:
        pass

    try:
        outputFile.write('<b>VBA</b><br><textarea rows="20" cols="85" wrap="OFF">')
        outputFile.write(pyeq3.outputSourceCodeService().GetOutputSourceCodeVBA(in_equation))
        outputFile.write('</textarea><br><br>\n\n')
    except:
        pass

    try:
        outputFile.write('<b>PYTHON</b><br><textarea rows="20" cols="85" wrap="OFF">')
        outputFile.write(pyeq3.outputSourceCodeService().GetOutputSourceCodePYTHON(in_equation))
        outputFile.write('</textarea><br><br>\n\n')
    except:
        pass

    try:
        outputFile.write('<b>JAVA</b><br><textarea rows="20" cols="85" wrap="OFF">')
        outputFile.write(pyeq3.outputSourceCodeService().GetOutputSourceCodeJAVA(in_equation))
        outputFile.write('</textarea><br><br>\n\n')
    except:
        pass

    try:
        outputFile.write('<b>JAVASCRIPT</b><br><textarea rows="20" cols="85" wrap="OFF">')
        outputFile.write(pyeq3.outputSourceCodeService().GetOutputSourceCodeJAVASCRIPT(in_equation))
        outputFile.write('</textarea><br><br>\n\n')
    except:
        pass

    try:
        outputFile.write('<b>JULIA</b><br><textarea rows="20" cols="85" wrap="OFF">')
        outputFile.write(pyeq3.outputSourceCodeService().GetOutputSourceCodeJULIA(in_equation))
        outputFile.write('</textarea><br><br>\n\n')
    except:
        pass

    try:
        outputFile.write('<b>FORTRAN90</b><br><textarea rows="20" cols="85" wrap="OFF">')
        outputFile.write(pyeq3.outputSourceCodeService().GetOutputSourceCodeFORTRAN90(in_equation))
        outputFile.write('</textarea><br><br>\n\n')
    except:
        pass

    try:
        outputFile.write('<b>SCILAB</b><br><textarea rows="20" cols="85" wrap="OFF">')
        outputFile.write(pyeq3.outputSourceCodeService().GetOutputSourceCodeSCILAB(in_equation))
        outputFile.write('</textarea><br><br>\n\n')
    except:
        pass

    try:
        outputFile.write('<b>MATLAB</b><br><textarea rows="20" cols="85" wrap="OFF">')
        outputFile.write(pyeq3.outputSourceCodeService().GetOutputSourceCodeMATLAB(in_equation))
        outputFile.write('</textarea><br><br>\n\n')
    except:
        pass

    outputFile.write('</body></html>\n')
    
    outputFile.close()


GraphUtils.py
import os, sys

import matplotlib
matplotlib.use('Agg') # immediately following the "import matplotlib" statement, web only

import matplotlib.pyplot as plt
import numpy, scipy, pyeq3


def SaveDataHistogram(in_filePath, in_data, in_title):
    
    fig = plt.figure(figsize=(5, 4))
    axes = fig.add_subplot(1,1,1)
    
    bincount = len(in_data) // 2
    if bincount < 5:
        bincount = 5
    if bincount > 25:
        bincount = 25
    n, bins, patches = axes.hist(in_data, bincount, rwidth=0.8)
    
    # some axis space at the top of the graph
    ylim = axes.get_ylim()
    if ylim[1] == max(n):
        axes.set_ylim(0.0, ylim[1] + 1)
    
    axes.set_title(in_title + ' Histogram') # add a title
    axes.set_xlabel(in_title) # X axis data label
    axes.set_ylabel(" Frequency") # Y axis label is always frequency
    
    plt.tight_layout() # prevents cropping axis labels
    fig.savefig(in_filePath) # create PNG file
    plt.close('all')
    

def SaveModelScatterConfidence(in_filePath, in_equation, in_title, in_xAxisLabel, in_yAxisLabel):
    
    # raw data
    x_data = in_equation.dataCache.allDataCacheDictionary['IndependentData'][0]
    y_data = in_equation.dataCache.allDataCacheDictionary['DependentData']
    
    # now create data for the fitted in_equation plot
    xModel = numpy.linspace(min(x_data), max(x_data))

    tempcache = in_equation.dataCache
    in_equation.dataCache = pyeq3.dataCache()
    in_equation.dataCache.allDataCacheDictionary['IndependentData'] = numpy.array([xModel, xModel])
    in_equation.dataCache.FindOrCreateAllDataCache(in_equation)
    yModel = in_equation.CalculateModelPredictions(in_equation.solvedCoefficients, in_equation.dataCache.allDataCacheDictionary)
    in_equation.dataCache = tempcache
    
    # now use matplotlib to create the PNG file
    fig = plt.figure(figsize=(5, 4))
    ax = fig.add_subplot(1,1,1)
    
    # first the raw data as a scatter plot
    ax.plot(x_data, y_data,  'D')
    
    # now the model as a line plot
    ax.plot(xModel, yModel)
    
    # now calculate confidence intervals
    # http://support.sas.com/documentation/cdl/en/statug/63347/HTML/default/viewer.htm#statug_nlin_sect026.htm
    # http://www.staff.ncl.ac.uk/tom.holderness/software/pythonlinearfit
    mean_x = numpy.mean(x_data)			# mean of x
    n = in_equation.nobs		    # number of samples in the origional fit
    
    t_value = scipy.stats.t.ppf(0.975, in_equation.df_e) # (1.0 - (a/2)) is used for two-sided t-test critical value, here a = 0.05
                            
    confs = t_value * numpy.sqrt((in_equation.sumOfSquaredErrors/in_equation.df_e)*(1.0/n + (numpy.power((xModel-mean_x),2.0)/
                                ((numpy.sum(numpy.power(x_data,2)))-n*(numpy.power(mean_x,2.0))))))

    # get lower and upper confidence limits based on predicted y and confidence intervals
    upper = yModel + abs(confs)
    lower = yModel - abs(confs)
    
    # mask off any numbers outside the existing plot limits
    booleanMask = yModel > matplotlib.pyplot.ylim()[0]
    booleanMask &= (yModel < matplotlib.pyplot.ylim()[1])
    
    # color scheme improves visibility on black background lines or points
    ax.plot(xModel[booleanMask], lower[booleanMask], linestyle='solid', color='white')
    ax.plot(xModel[booleanMask], upper[booleanMask], linestyle='solid', color='white')
    ax.plot(xModel[booleanMask], lower[booleanMask], linestyle='dashed', color='blue')
    ax.plot(xModel[booleanMask], upper[booleanMask], linestyle='dashed', color='blue')
    
    ax.set_title(in_title) # add a title
    ax.set_xlabel(in_xAxisLabel) # X axis data label
    ax.set_ylabel(in_yAxisLabel) # Y axis data label
    
    plt.tight_layout() # prevents cropping axis labels
    fig.savefig(in_filePath) # create PNG file
    plt.close('all')


def SurfaceAndContourPlots(in_filePathSurface, in_filePathContour, in_equation,
                           in_surfaceTitle, in_contourTitle,
                           in_xAxisLabel, in_yAxisLabel, in_zAxisLabel):

    # raw data
    x_data = in_equation.dataCache.allDataCacheDictionary['IndependentData'][0]
    y_data = in_equation.dataCache.allDataCacheDictionary['IndependentData'][1]
    z_data = in_equation.dataCache.allDataCacheDictionary['DependentData']

    from mpl_toolkits.mplot3d import Axes3D # 3D apecific
    from matplotlib import cm # to colormap from blue to red
    
    fig = plt.figure()
    ax = fig.gca(projection='3d')

    xModel = numpy.linspace(min(x_data), max(x_data), 20)
    yModel = numpy.linspace(min(y_data), max(y_data), 20)
    X, Y = numpy.meshgrid(xModel, yModel)

    tempcache = in_equation.dataCache
    in_equation.dataCache = pyeq3.dataCache()
    in_equation.dataCache.allDataCacheDictionary['IndependentData'] = numpy.array([X, Y])
    in_equation.dataCache.FindOrCreateAllDataCache(in_equation)
    Z = in_equation.CalculateModelPredictions(in_equation.solvedCoefficients, in_equation.dataCache.allDataCacheDictionary)
    in_equation.dataCache = tempcache
    
    ax.plot_surface(X, Y, Z, rstride=1, cstride=1, cmap=cm.coolwarm,
            linewidth=1, antialiased=True)
    
    ax.scatter(x_data, y_data, z_data)
    
    ax.set_title(in_surfaceTitle) # add a title for surface plot
    ax.set_xlabel(in_xAxisLabel) # X axis data label
    ax.set_ylabel(in_yAxisLabel) # Y axis data label
    ax.set_zlabel(in_zAxisLabel) # Y axis data label
    
    plt.tight_layout() # prevents cropping axis labels
    fig.savefig(in_filePathSurface) # create PNG file
    plt.close('all')

    # contour plot here
    fig = plt.figure(figsize=(5, 4))
    ax = fig.add_subplot(1,1,1)

    ax.plot(x_data, y_data, 'o', color='0.8', markersize=4) # draw these first so contour lines overlay.  Color=number is grayscale

    ax.set_title(in_contourTitle) # add a title for contour plot
    ax.set_xlabel(in_xAxisLabel) # X data label
    ax.set_ylabel(in_yAxisLabel) # Y data label

    numberOfContourLines = 16
    CS = plt.contour(X, Y, Z, numberOfContourLines, colors='k')
    plt.clabel(CS, inline=1, fontsize=10) # labels for contours

    plt.tight_layout() # prevents cropping axis labels
    fig.savefig(in_filePathContour) # create PNG file
    plt.close('all')


def SaveAbsErrorScatterPlot(in_filePath, in_equation, in_title, in_xAxisLabel):
    
    # raw data
    dep_data = in_equation.dataCache.allDataCacheDictionary['DependentData']
    abs_err = in_equation.modelAbsoluteError

    # now use matplotlib to create the PNG file
    fig = plt.figure(figsize=(5, 4))
    ax = fig.add_subplot(1,1,1)
    
    # create the scatter plot
    ax.plot(dep_data, abs_err, 'D')
    
    ax.set_title(in_title) # add a title
    ax.set_xlabel(in_xAxisLabel) # X axis data label
    ax.set_ylabel(" Absolute Error") # Y axis label is always is absolute error

    plt.tight_layout() # prevents cropping axis labels
    fig.savefig(in_filePath) # create PNG file
    plt.close('all')


def SavePercentErrorScatterPlot(in_filePath, in_equation, in_title, in_xAxisLabel):
    
    # raw data
    dep_data = in_equation.dataCache.allDataCacheDictionary['DependentData']
    percent_err = in_equation.modelPercentError

    # now use matplotlib to create the PNG file
    fig = plt.figure(figsize=(5, 4))
    ax = fig.add_subplot(1,1,1)
    
    # create the scatter plot
    ax.plot(dep_data, percent_err, 'D')
    
    ax.set_title(in_title) # add a title
    ax.set_xlabel(in_xAxisLabel) # X axis data label
    ax.set_ylabel(" Percent Error") # Y axis label is always is percent error

    plt.tight_layout() # prevents cropping axis labels
    fig.savefig(in_filePath) # create PNG file
    plt.close('all')


FlaskFit.py
import os, sys, inspect

import pyeq3, GraphUtils, TextUtils, UserInterface
from flask import Flask
from flask import request


# override Flask's default file cache for the files we generate
class MyFlask(Flask):
    def get_send_file_max_age(self, name):
        if name.lower().endswith('.png'):
            return 0.000001
        if name.lower().endswith('.txt'):
            return 0.000001
        if name.lower().endswith('.html'):
            return 0.000001
        return Flask.get_send_file_max_age(self, name)


app = MyFlask(__name__)
app.debug = True # only for development, never for production


@app.route('/')
def UserInterfaceHTML():
    # return HTML to Flask as a web page
    s = '<html><body>'
    s += '<table><tr>'
    s += '<td>' + UserInterface.htmlForm_2D + '</td>'
    s += '<td> </td>'
    s += '<td>' + UserInterface.htmlForm_3D + '</td>'
    s += '</tr></table>'
    s +='</body></html>'

    return s


@app.route('/simplefitter_2D', methods=['POST'])
def simplefitter_2D_NoFormDataValidation():
    formTextData = request.form['textdata']
    formEquation = request.form['equation']
    formFittingTarget = request.form['target']

    if formEquation == 'Linear':
        equation = pyeq3.Models_2D.Polynomial.Linear(formFittingTarget)
    elif formEquation == 'Quadratic':
        equation = pyeq3.Models_2D.Polynomial.Quadratic(formFittingTarget)
    elif formEquation == 'Cubic':
        equation = pyeq3.Models_2D.Polynomial.Cubic(formFittingTarget)
    elif formEquation == 'WitchA':
        equation = pyeq3.Models_2D.Miscellaneous.WitchOfAgnesiA(formFittingTarget)
    elif formEquation == 'LorentzianPeakCWithOffset':
        equation = pyeq3.Models_2D.Peak.LorentzianPeakC(formFittingTarget, 'Offset')
    elif formEquation == 'GammaRayDegreesB':
        equation = pyeq3.Models_2D.LegendrePolynomial.GammaRayAngularDistributionDegreesB(formFittingTarget)
    elif formEquation == 'ExponentialWithOffset':
        equation = pyeq3.Models_2D.Exponential.Exponential(formFittingTarget, 'Offset')

    # the name of the data here is from the form
    # check for functions requiring non-zero nor non-negative data such as 1/x, etc.
    try:
        pyeq3.dataConvertorService().ConvertAndSortColumnarASCII(formTextData, equation, False)
    except:
        return equation.reasonWhyDataRejected

    # check for number of coefficients > number of data points to be fitted
    coeffCount = len(equation.GetCoefficientDesignators())
    dataCount = len(equation.dataCache.allDataCacheDictionary['DependentData'])
    if coeffCount > dataCount:
        return "This equation requires a minimum of " + repr(coeffCount) + " data points, you supplied " + repr(dataCount) + "."

    equation.Solve()
    equation.CalculateModelErrors(equation.solvedCoefficients, equation.dataCache.allDataCacheDictionary)
    equation.CalculateCoefficientAndFitStatistics()

    # save fit statistics to a text file
    fitStatisticsFilePath = "static/fitstatistics_2D.txt"
    TextUtils.SaveCoefficientAndFitStatistics(fitStatisticsFilePath,  equation)

    # save source code to a single text file, all available languages
    sourceCodeFilePath = "static/sourcecode_2D.html"
    TextUtils.SaveSourceCode(sourceCodeFilePath,  equation)

    # create graph
    graphFilePath = "static/model_and_scatterplot_2D.png"
    title = "Model with 95% Confidence Intervals"
    xAxisLabel = "X data"
    yAxisLabel = "Y data"
    GraphUtils.SaveModelScatterConfidence(graphFilePath,
                                          equation, title, xAxisLabel, yAxisLabel) 

    absErrorPlotFilePath = "static/abs_error_2D.png"
    title = "Absolute Error"
    GraphUtils.SaveAbsErrorScatterPlot(absErrorPlotFilePath, equation, title, yAxisLabel)
    
    absErrorHistFilePath = "static/abs_error_hist_2D.png"
    title = "Absolute Error"
    GraphUtils.SaveDataHistogram(absErrorHistFilePath, equation.modelAbsoluteError, title)
    
    if equation.dataCache.DependentDataContainsZeroFlag != 1:
        percentErrorPlotFilePath = "static/per_error_2D.png"
        title = "Percent Error"
        GraphUtils.SavePercentErrorScatterPlot(percentErrorPlotFilePath, equation, title, yAxisLabel)
        
        perErrorHistFilePath = "static/per_error_hist_2D.png"
        title = "Percent Error"
        GraphUtils.SaveDataHistogram(perErrorHistFilePath, equation.modelPercentError, title)


    # generate HTML
    htmlToReturn = ''
    htmlToReturn +=  equation.GetDisplayName() + '<br><br>\n'
    htmlToReturn +=  equation.GetDisplayHTML() + '<br><br>\n'
    htmlToReturn += '<a href="' + fitStatisticsFilePath + '">Link to parameter and fit statistics</a><br><br>\n'
    htmlToReturn += '<a href="' + sourceCodeFilePath + '">Link to source code, all available languages</a><br><br>\n'
    htmlToReturn += '<a href="static/AdditionalInfo.html">Link to additional information</a><br><br>\n'
    htmlToReturn +=  '<img src="' + graphFilePath + '"> <br>\n'
    htmlToReturn +=  '<img src="' + absErrorPlotFilePath + '"><br>\n'
    htmlToReturn +=  '<img src="' + absErrorHistFilePath + '"><br>\n'
    
    if equation.dataCache.DependentDataContainsZeroFlag != 1:
        htmlToReturn +=  '<img src="' + percentErrorPlotFilePath + '"><br><br>\n'
        htmlToReturn +=  '<img src="' + perErrorHistFilePath + '"><br><br>\n'

    return '<html><body>' + htmlToReturn + '</body></html>'


@app.route('/simplefitter_3D', methods=['POST'])
def simplefitter_3D_NoFormDataValidation():
    
    formTextData = request.form['textdata']
    formEquation = request.form['equation']
    formFittingTarget = request.form['target']

    if formEquation == 'Linear':
        equation = pyeq3.Models_3D.Polynomial.Linear(formFittingTarget)
    elif formEquation == 'FullQuadratic':
        equation = pyeq3.Models_3D.Polynomial.FullQuadratic(formFittingTarget)
    elif formEquation == 'FullCubic':
        equation = pyeq3.Models_3D.Polynomial.FullCubic(formFittingTarget)
    elif formEquation == 'MonkeySaddleA':
        equation = pyeq3.Models_3D.Miscellaneous.MonkeySaddleA(formFittingTarget)
    elif formEquation == 'GaussianCurvatureOfWhitneysUmbrellaA':
        equation = pyeq3.Models_3D.Miscellaneous.GaussianCurvatureOfWhitneysUmbrellaA(formFittingTarget)
    elif formEquation == 'NIST_NelsonAutolog':
        equation = pyeq3.Models_3D.NIST.NIST_NelsonAutolog(formFittingTarget)
    elif formEquation == 'CustomPolynomialOne': # X order 3, Y order 1 in this example - passed as integers
        equation = pyeq3.Models_3D.Polynomial.UserSelectablePolynomial(formFittingTarget, "Default", 3, 1)
    
    # the name of the data here is from the form
    # check for functions requiring non-zero nor non-negative data such as 1/x, etc.
    try:
        pyeq3.dataConvertorService().ConvertAndSortColumnarASCII(formTextData, equation, False)
    except:
        return equation.reasonWhyDataRejected

    # check for number of coefficients > number of data points to be fitted
    coeffCount = len(equation.GetCoefficientDesignators())
    dataCount = len(equation.dataCache.allDataCacheDictionary['DependentData'])
    if coeffCount > dataCount:
        return "This equation requires a minimum of " + repr(coeffCount) + " data points, you supplied " + repr(dataCount) + "."

    equation.Solve()
    equation.CalculateModelErrors(equation.solvedCoefficients, equation.dataCache.allDataCacheDictionary)
    equation.CalculateCoefficientAndFitStatistics()

    # save fit statistics to a text file
    fitStatisticsFilePath = "static/fitstatistics_3D.txt"
    TextUtils.SaveCoefficientAndFitStatistics(fitStatisticsFilePath,  equation)

    # save source code to a single text file, all available languages
    sourceCodeFilePath = "static/sourcecode_3D.html"
    TextUtils.SaveSourceCode(sourceCodeFilePath,  equation)

    # create graphs
    graphFilePath_Surface = "static/surface.png" # surface plot
    graphFilePath_Contour = "static/contour.png" # contour plot
    surfaceTitle = "Surface Plot"
    contourTitle = "Contour Plot"
    xAxisLabel = "X data"
    yAxisLabel = "Y data"
    zAxisLabel = "Z data"
    GraphUtils.SurfaceAndContourPlots(graphFilePath_Surface,
                                      graphFilePath_Contour,
                                      equation, surfaceTitle, contourTitle,
                                      xAxisLabel, yAxisLabel, zAxisLabel)

    absErrorPlotFilePath = "static/abs_error_3D.png"
    title = "Absolute Error"
    GraphUtils.SaveAbsErrorScatterPlot(absErrorPlotFilePath, equation, title, zAxisLabel)

    absErrorHistFilePath = "static/abs_error_hist_3D.png"
    title = "Absolute Error"    
    GraphUtils.SaveDataHistogram(absErrorHistFilePath, equation.modelAbsoluteError, title)

    if equation.dataCache.DependentDataContainsZeroFlag != 1:
        perErrorPlotFilePath = "static/per_error_3D.png"
        title = "Percent Error"
        GraphUtils.SavePercentErrorScatterPlot(perErrorPlotFilePath, equation, title, zAxisLabel)
        
        perErrorHistFilePath = "static/per_error_hist_3D.png"
        title = "Percent Error"
        GraphUtils.SaveDataHistogram(perErrorHistFilePath, equation.modelPercentError, title)

    # generate HTML
    htmlToReturn = ''
    htmlToReturn +=  equation.GetDisplayName() + '<br><br>\n'
    htmlToReturn +=  equation.GetDisplayHTML() + '<br><br>\n'
    htmlToReturn += '<a href="' + fitStatisticsFilePath + '">Link to parameter and fit statistics</a><br><br>\n'
    htmlToReturn += '<a href="' + sourceCodeFilePath + '">Link to source code, all available languages</a><br><br>\n'
    htmlToReturn += '<a href="static/AdditionalInfo.html">Link to additional information</a><br><br>\n'
    htmlToReturn +=  '<img src="' + graphFilePath_Surface + '"><br><br>\n'
    htmlToReturn +=  '<img src="' + graphFilePath_Contour + '"><br><br>\n'
    htmlToReturn +=  '<img src="' + absErrorPlotFilePath + '"><br><br>\n'
    htmlToReturn +=  '<img src="' + absErrorHistFilePath + '"><br><br>\n'
    if equation.dataCache.DependentDataContainsZeroFlag != 1:
        htmlToReturn +=  '<img src="' + perErrorPlotFilePath + '"><br><br>\n'
        htmlToReturn +=  '<img src="' + perErrorHistFilePath + '"><br><br>\n'

    return '<html><body>' + htmlToReturn + '</body></html>'


@app.route('/equationlist_2D', methods=['GET'])
def equationlist_2D():
    htmlToReturn = '' # build this as we progress

    htmlToReturn += '<table border=1>'
    
    for submodule in inspect.getmembers(pyeq3.Models_2D):
        if inspect.ismodule(submodule[1]):
            for equationClass in inspect.getmembers(submodule[1]):
                if inspect.isclass(equationClass[1]):
                    for extendedVersionName in ['Default', 'Offset']:
                        if (-1 != extendedVersionName.find('Offset')) and (equationClass[1].autoGenerateOffsetForm == False):
                            continue
    
                        equation = equationClass[1]('SSQABS', extendedVersionName)
                        htmlToReturn += '<tr>'
                        htmlToReturn += '<td nowrap><b>2D ' + submodule[0] + '</b></td>'
                        htmlToReturn += '<td nowrap><i>' + equation.GetDisplayName() + '</i></td>'
                        htmlToReturn += '<td nowrap>' + equation.GetDisplayHTML() + '</td>'
                        htmlToReturn += '</tr>'
                        
    htmlToReturn += '</table>'
                        
    return '<html><body>' + htmlToReturn + '</body></html>'


@app.route('/equationlist_3D', methods=['GET'])
def equationlist_3D():
    htmlToReturn = '' # build this as we progress
    
    htmlToReturn += '<table border=1>'
    
    for submodule in inspect.getmembers(pyeq3.Models_3D):
        if inspect.ismodule(submodule[1]):
            for equationClass in inspect.getmembers(submodule[1]):
                if inspect.isclass(equationClass[1]):
                    for extendedVersionName in ['Default', 'Offset']:
                        if (-1 != extendedVersionName.find('Offset')) and (equationClass[1].autoGenerateOffsetForm == False):
                            continue
    
                        equation = equationClass[1]('SSQABS', extendedVersionName)
                        htmlToReturn += '<tr>'
                        htmlToReturn += '<td nowrap><b>3D ' + submodule[0] + '</b></td>'
                        htmlToReturn += '<td nowrap><i>' + equation.GetDisplayName() + '</i></td>'
                        htmlToReturn += '<td nowrap>' + equation.GetDisplayHTML() + '</td>'
                        htmlToReturn += '</tr>'
                        
    htmlToReturn += '</table>'

    return '<html><body>' + htmlToReturn + '</body></html>'


if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000)
